### PR TITLE
Fix `RNN`s backward with `zoneout`

### LINKED ISSUE #43: `RNN`s with `zoneout > 0.0` have wrong gradients
`RNN`s use the zoneout mask of the first step to compute gradients for all steps.

For example, maybe we could change `GRU` backward from
https://github.com/lmnt-com/haste/blob/3d92d702fa68a6705444eaabc5aa873549c54c05/lib/gru_backward_gpu.cu.cc#L371
to
```python
zoneout_mask ? zoneout_mask + i * NH : nullptr
```

### PR DESCRIPTION
**Pull request by manipopopo**
Fix #43.

### CODE REVIEW

File: validation/pytorch.py [hunk 1]
```diff
@@ -40,16 +41,21 @@
 
 def self_consistency(rnn, x):
   x_cuda = x.clone().cuda()
-  x_cpu = x.clone().cpu()
+  x_cuda_torch = x_cuda.detach().clone()
```
Comment (sharvil) [inline validation/pytorch.py:44]: I'm curious why you chose to mock `_is_cuda` on the RNN instead of passing in a CPU tensor directly. Was there an issue with the previous approach?

File: validation/pytorch.py [hunk 2]
```diff
@@ -40,16 +41,21 @@
...
   x_cuda.requires_grad_(True)
-  x_cpu.requires_grad_(True)
+  x_cuda_torch.requires_grad_(True)
 
-  y1, _ = rnn.cuda().forward(x_cuda)
+  rnn.cuda()
+  seed = 5566
+  torch.manual_seed(seed)
```
Comment (sharvil) [inline validation/pytorch.py:50]: It's probably a good idea to control the seed for reproducible test results. Maybe not hard-coded in here like this, though? I suggest either removing the RNG seeding, or setting the RNG seed through a command line argument with a default of no-explicit-seed.

**Review by sharvil — COMMENTED** [review]
Overall this code looks great. Thanks so much – both for the bugreport and the fix! I've added a couple of comments, and am happy to merge when resolved.

File: validation/pytorch.py (hunk 1 above)
Comment (manipopopo) [reply validation/pytorch.py:44]: [Currently *PyTorch* cannot get identical random masks](https://pytorch.org/docs/1.10.1/notes/randomness.html) (*zoneout* masks or *dropout* masks) on `"cpu"` and `"cuda"`. So we compare _PyTorch_ version *RNN*s on `"cuda"` with haste *RNN*s, which only have GPU implementation.

To run the _PyTorch_ version *RNN*s on `"cuda"`, maybe we could create RNNs like:

```python
pytorch_rnn = haste.GRU(32, 256, backend="pytorch").cuda()
haste_rnn = haste.GRU(32, 256, backend="haste").cuda()
```

But the changes of interfaces seems to be off topic. It seems to me that mocking `_is_cuda` is the simplest way to run *PyTorch* version `RNN`s on `"cuda"`.

File: validation/pytorch.py (hunk 2 above)
Comment (manipopopo) [reply validation/pytorch.py:50]: We set the seed manually to make sure that the two runs will use identical random masks (*zoneout* masks and *dropout* masks) generated inside `forward`. Maybe the seed could be some random value hard-coded like [pytorch tests](https://github.com/pytorch/pytorch/blob/20489ebdc92068823941de285474f9b5a745c68d/test/test_cuda.py#L656-L664)?

Comment (sharvil) [reply validation/pytorch.py:50]: Ah, this makes sense – you're re-seeding between the two runs to generate the same random masks, and not to just get reproducible results across multiple runs of this script.

File: validation/pytorch.py (hunk 1 above)
Comment (sharvil) [reply validation/pytorch.py:44]: I see, thanks for clarifying. Your approach looks right.

**Review by sharvil — APPROVED** [review]
